tensorboard/webapp/widgets/custom_modal/custom_modal.ts

```diff
@@ -0,0 +1,221 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {
+  ApplicationRef,
+  Injectable,
+  TemplateRef,
+  ViewContainerRef,
+} from '@angular/core';
+import {ConnectedPosition, Overlay, OverlayRef} from '@angular/cdk/overlay';
+import {TemplatePortal} from '@angular/cdk/portal';
+import {Subscription} from 'rxjs';
+import {isMouseEventInElement} from '../../util/dom';
+
+/**
+ * Enables dynamic creation of modal components.
+ *
+ * # Prerequisites
+ * App root component must define a ViewContainerRef named `vcRef`
+ * e.g.:
+ * ```
+ * // my_app_root.ts
+ * constructor(readonly vcRef: ViewContainerRef) {}
+ * ```
+ *
+ * # Usage
+ * Define a modal using an ng-template:
+ * ```
+ * <ng-template #myModalTemplate>
+ *    <my-awesome-modal-content
+ *      [input1]="input1"
+ *      (output1)="output1"
+ *    >
+ *    </my-awesome-modal-content>
+ * </ng-template>
+ * ```
+ *
+ * Define a ViewChild to reference the template in the component file:
+ * ```
+ * // my_component.ts
+ * ...
+ * @ViewChild('myModalTemplate', {read: TemplateRef})
+ * myModalTemplate!: TemplateRef<unknown>;
+ * ...
+ * ```
+ *
+ * Inject CustomModal into the component
+ * ```
+ * // my_component.ts
+ * ...
+ * constructor(private readonly customModal: CustomModal) {}
+ * ...
+ * ```
+ *
+ * To create a modal, call createNextToElement():
+ * ```
+ * // my_component.ts
+ * ...
+ * onSomeButtonClick(mouseEvent: MouseEventj) {
+ *   this.customModal.createNextToElement(this.myModalTemplate, mouseEvent.target as HTMLElement);
+ * }
+ * ...
+ * ```
+ *
+ * # Testing
+ * A similar setup is required for testing modal behavior from another component.
+ *
+ * Define a separate root component to hold the modal container and ref:
+ * ```
+ * @Component({
+ *   ...,
+ *   template: `<actual-component-to-test></actual-component-to-test>`
+ * })
+ * class TestableComponent {
+ *   constructor(readonly vcRef: ViewContainerRef) {}
+ *   ...
+ * }
+ * ```
+ *
+ * Before each test (e.g. in beforeEach), make sure to add the TestableComponent
+ * to the app component list:
+ * ```
+ * ...
+ * const fixture = TestBed.createComponent(TestableComponent);
+ * const appRef = TestBed.inject(ApplicationRef);
+ * appRef.components.push(fixture.componentRef);
+ * ```
+ */
+
+export class CustomModalRef {
+  overlayRef: OverlayRef;
+  subscriptions: Subscription[];
+
+  constructor(overlayRef: OverlayRef, subscriptions: Subscription[] = []) {
```

Comment:
As far as I can tell you never override subscriptions with something else. Can you just define its initial `[]` value on L103?

Reply:
Oh right, I'm appending the subscriptions one by one after instantiation. Changed as described and added simple tests to check subscribe/unsubscribe logic